File: R/FireweedUtils.r
#FireweedUtils.r
#Programmed by: Joshua M. Rady
#Woodwell Climate Research Center
#2024
#Reference: Proj. 11 Exp. 19
#
#Description:---------------------------------------------------------------------------------------
#  This file is part of the Fireweed wildfire code library.
#  This file contains (general) utilities.
#
#___________________________________________________________________________________________________

#This utility checks that the parameters (vectors) passed have the same length.  Between 2 and 4
#arguments are accepted.
#Adapted from code originally in in CalcWeightings().
#
#Alternatively we could return the length if true and otherwise -1, but this seems likely to create
#more work in practice.  In C positive = TRUE, 0 = FALSE, which could be more useful.
#In the current usage we expect the arguments to be a mix of numeric and logical vectors.  We
#could add checking for this.
SameLengths <- function(arg1, arg2, arg3 = NULL, arg4 = NULL)
{
  #Put the argments in a list removing NULL elements:
  argList = list(arg1, arg2, arg3, arg4)
  argList = argList[!sapply(argList, is.null)]
  #This would work too for omitted arguments but would ignore any zero length vectors passed in:
  #argList = argList[length(argList) != 0]
  
  #Are arguments the same length?
  if (all(sapply(argList, length) == length(arg1)))
  {
    #return(length(arg1))
    return(TRUE)
  }
  else
  {
    #return(-1)
    return(FALSE)
  }
}

#This utility checks that a value falls in a valid range.
#
#R array aware.
InRange <- function(value, low, high)
{
  if (any(value < low) || any(value > high))
  {
    return(FALSE)
  }
  else
  {
    return(TRUE)
  }
}

#Check if a value is from 0 to 1, a valid proportion.
#
#R array aware.
ValidProportion <- function(value)
{
  return(InRange(value, low = 0, high = 1))
}

File: C++/FireweedUtils.cpp
/***************************************************************************************************
FireweedUtils.cpp
Programmed by: Joshua M. Rady
Woodwell Climate Research Center
Started: 8/12/2024
Reference: Proj. 11 Exp. 19

	This file is part of the Fireweed wildfire code library.
	This file defines a set of (general) utilities.

***************************************************************************************************/

#include <cmath>//fabs()

#include "FireweedUtils.h"

//See the header for SameLengths().

//This utility checks that a value falls in a valid range.
bool InRange(double value, double low, double high)
{
	return (value >= low && value <= high);
}

bool InRange(std::vector<double> value, double low, double high)
{
	for (int i = 0; i < value.size(); i++)
	{
		if (value[i] < low || value[i] > high)
		{
			return false;
		}
	}

	return true;
}

/** Check if a value is from 0 to 1, a valid proportion.
 *
 * @param value A number to check.
 *
 * @returns Is the value from 0 to 1.
 */
bool ValidProportion(double value)
{
	return InRange(value, 0, 1);
}

/** Check if a value is from 0 to 1, a valid proportion.
 *
 * @param value A vector of numbers to check.
 *
 * @returns Are all the values from 0 to 1.
 */
bool ValidProportion(std::vector<double> value)
{
	return InRange(value, 0, 1);
}

/** Compare two floating point values for near/effective equality:
 *
 * It is easy for math operations to result in small floating point errors that may make two
 * variables you expect to be identical to be slightly different.  Use this fucntion in place of ==
 * in such cases.
 *
 * @param val1 The first value to compare.
 * @param val2 The second value to compare.
 * @param precision How close must the values be to be "equal".
 *
 * @returns True if the two values are nearly equal.
 *
 * @note I'm not sure what the default should be for the precision of this comparison (see header for default value).
 * @note This function is currently C++ only.
 */
bool FloatCompare(double val1, double val2, double precision)//Or epsilon?
{
	if (std::fabs(val1 - val2) < precision)
	{
		return true;
	}
	else
	{
		return false;
	}
}
